lms/djangoapps/badges/backends/badgr.py

```diff
@@ -166,12 +166,61 @@ def _create_assertion(self, badge_class, user, evidence_url):
         self._send_assertion_created_event(user, assertion)
         return assertion
 
-    @staticmethod
-    def _get_headers():
+    def _get_and_cache_oauth_tokens(self, refresh_token=None):
+        data = {
+            'username': settings.BADGR_USERNAME,
+            'password': settings.BADGR_PASSWORD,
+        }
+        if refresh_token:
+            data = {
+                'grant_type': 'refresh_token',
+                'refresh_token': refresh_token
+            }
+
+        oauth_url = "{}/o/token".format(settings.BADGR_BASE_URL)
+
+        response = requests.post(
+            oauth_url, data=data, timeout=settings.BADGR_TIMEOUT
+        )
+        response.raise_for_status()
+        try:
+            data = response.json()
+            # set the access token expiry 5 seconds early to make sure
+            # it will not expire while in use
+            expires_in = data['expires_in'] - 5
+            result = {
+                'access_token': data['access_token'],
+                'refresh_token': data['refresh_token'],
+                'expires_at': datetime.datetime.utcnow() + datetime.timedelta(
+                    seconds=expires_in)
+            }
+            TieredCache.set_all_tiers(
+                settings.BADGR_TOKENS_CACHE_KEY, result, expires_in)
+            return result
+        except (KeyError, json.decoder.JSONDecodeError) as json_error:
+            raise requests.RequestException(response=response) from json_error
+
+    def _get_access_token(self):
+        cached_response = TieredCache.get_cached_response(
+            settings.BADGR_TOKENS_CACHE_KEY)
+        if cached_response.is_found:
+            cached_tokens = cached_response.value
+            if cached_tokens.get('expires_at') > datetime.datetime.utcnow():
+                return cached_tokens.get('access_token')
+            else:
+                refreshed_tokens = self._get_and_cache_oauth_tokens(
+                    refresh_token=cached_tokens.get('refresh_token'))
```

Comment:
This `else` block is a "can't happen" situation, right? Meaning, the cache should never return an expired token. Thus, if `is_found` returns `True` **and** the expiry is in the past, I'd suggest to perhaps add a big screaming log message, because in that case I'd argue that something is very wrong with the cache implementation.

Reply:
So the cache holds an `access_token`, `access_token` expiry timestamp and a `refresh_token`, it could return an expired `access_token` and then it would be the time to go and renew it with the `refresh_token`. At least this was my initial thought, and it works, but I am not sure what the cache expiry should be then.